### PR TITLE
#432 Invite users using an excel file

### PR DESCRIPTION
**Pull request by Oliviier-dev**
### PR Description
Admin can invite many users to join the organization by uploading an excel file.
### Description of tasks that were expected to be completed

- Allow admin to upload an excel file.
- Retrieve the valid data from the excel(user email and role).
- Invite all the valid users.

### Functionality
This functionality allows an admin to upload an excel file with email and role columns, where all the rows with valid emails and role, will be sent an email inviting them to register to the organization
### How has this been tested?
Clone this repository and navigate into it. Open the project in your text editor. Run npm install to install the dependencies, followed by npm run seed to seed the database, and npm run dev to start the server. Afterward, log in as the admin and use the UploadInvitationFile mutation with an Excel file containing the users you want to invite.
### PR Checklist:
- [x] Task 1. Allow admin to upload an excel file.
- [x] Task 2. Retrieve the valid data from the excel(user email and role).
- [x] Task 3. Invite all the valid users via email.
### Track PR
Trello Link [(#432)](https://github.com/atlp-rwanda/atlp-pulse-fn/issues/432)

#### The uploaded excel file
Screenshot 2024-09-03 at 14 50 10<br>
#### The mutation which sends a request to upload an invitation file
Screenshot 2024-09-03 at 14 53 37<br>
#### The invitations created in the database with their respective roles
Screenshot 2024-09-03 at 14 54 00<br>
#### The email sent to the invited users
Screenshot 2024-09-04 at 09 26 18

### CODE REVIEW

**Review by Yvan-David — COMMENTED** [review]
Great Job

Please add some screenshots

1. shows that the invitation email have been sent to users
2. after signup the users will have the roles given in excel not the default role(you may use the db to show it)

**Review by niyobertin — APPROVED** [review]

**Review by Yvan-David — APPROVED** [review]

**Review by YvetteNyibuka — APPROVED** [review]

File: src/resolvers/invitation.resolvers.ts [hunk 1]
```diff
@@ -73,6 +97,136 @@ const invitationResolvers: IResolvers = {
         throw new ApolloError(error.message, 'INTERNAL_SERVER_ERROR');
       }
     },
+
+    async uploadInvitationFile(
```
Comment (ncutixavier) [inline src/resolvers/invitation.resolvers.ts:88]: Can you try refactoring this function? it is too long

**Review by niyobertin — APPROVED** [review]

File: src/resolvers/invitation.resolvers.ts [hunk 2]
```diff
@@ -10,7 +13,120 @@ import inviteUserTemplate from '../utils/templates/inviteUserTemplate';
...
+} as const
+
+type Role = typeof ROLE[keyof typeof ROLE]
+
+function isValidRole(role: string): boolean {
+  return (
+    typeof role === 'string' &&
+    Object.values(ROLE).includes(role.toLowerCase() as Role)
+  )
+}
```
Comment (ncutixavier) [inline src/resolvers/invitation.resolvers.ts]: Validation should not be here

File: src/resolvers/invitation.resolvers.ts [hunk 3]
```diff
@@ -10,7 +13,120 @@ import inviteUserTemplate from '../utils/templates/inviteUserTemplate';
...
+    const invitees: { email: string; role: Role }[] = []
+    const invalidRows: string[] = []
+
+    workbook.SheetNames.forEach((sheetName) => {
+      const worksheet: xlsx.WorkSheet = workbook.Sheets[sheetName]
+      const data: any[] = xlsx.utils.sheet_to_json(worksheet)
+
+      data.forEach((row) => {
+        const email = (row as { email?: string }).email?.trim()
+        const role = (row as { role?: string }).role?.trim().toLowerCase()
```
Comment (ncutixavier) [inline src/resolvers/invitation.resolvers.ts]: This should not be here

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**